### ISSUE TITLE
Formula seems to have issue managing results for more than ~30 containers

### ISSUE THREAD

**Issue by dsaingre** [powerapi-ng/smartwatts-formula#21]
## My issue
I am trying to use PowerAPI to manage the energy consumption of a micro-services based application. The [application](https://github.com/delimitrou/DeathStarBench/tree/master/socialNetwork) in itself is composed of ~30 containers. My issue is that the formula seems to be to slow to compute the power consumption of those containers : there is a growing lag between the current time of the experiment and the timestamp of the last value generated by the formula. 

I'm using the last docker version of HWPC-sensor and Smartwatt formula. Everything apart from grafana is deployed on a Ubuntu 20.04 server. 

## Investigations 
This issue is manifested by a growing number of values in the MongoDB Sensor collection. From what I understand, this number should stay low as the formula delete each value it reads.

### Trying to see if the issue come from the number of monitored containers
I tried to isolate the issue by deploying my power api (mongo, sensor, formula) and monitoring stack (influxdb as formula output and Grafana). When deploying a growing number of dumb containers (nginx), the issue seems to appear when there is around 25-30 containers to monitor. The sensor seems to do just fine but the formula seems to can't keep up with the data volume generated by the sensor. 

### Trying to figure out where the issue come from 
I believe the issue is not coming from the time needed by the formula to compute the energy estimation. The issue seems to be the interaction with the mongo database (reading and deleting the values generated by the sensor). Here is a flame graph of the formula process activity. 

![graph](https://user-images.githubusercontent.com/7606929/153576019-f9e1ffe3-5138-4739-93f3-0db298cdde46.svg)

If I read it correctly, interaction with mongodb (via pymongo) takes up for 80% of the formula activity. Calls to the `compute_power_estimation` of the formula takes up for 0.13% of the time and interaction with influxdb takes up for less than 4% of the formula activity. 

## Potential solutions ?
My guess is that, in stream mode, the mongo client (in powerapi) is reading and deleting values one by one. Such a high number of request could explain the amount of time spent by pymongo in the overall formula activity. 

Maybe bulk read and delete could be considered to limit database interactions (not sure on this part).

Does someone else have the same scalability issue ? If so, may the issue come from the db interactions?

**Comment by dsaingre** [comment]
Not sure if the attachment can be downloaded. I can send it directly if needed.

**Comment by PierreRustOrange** [comment]
Hi,

I also experience this type of lag when running even a moderate number of containers (e.g a dozen), however I'm not using mongodb but a straight socket connection between the sensor and the formula, so I think there is also an issue with the way the formula handles the report.

As a workaround, you can increase the polling period on the sensor and the formula in order to generate less reports.

**Comment by MadalinaDinga** [comment]
I have the same issue when running it on an MSA with 41 microservices. I use MongoDb both for pulling the HWPC reports and for pushing the power measurements.

**Comment by msalvati1997** [comment]
Hi,
I have the same issue.
I use mongoDB for HWPC reports and InfluxDB for SMARTWATTS reports.

Have you found a solution?

**Comment by roda82** [comment]
Hi, 
Which version of smartwatts are you using ?

**Comment by msalvati1997** [comment]
Thank you for your prompt reply.

I use the latest smartwatts version in docker-hub.
What I have done is to build a docker-compose that puts up the services I need to use  : 
```
version: '3.4'

services:

  influxdb:
    container_name: influx
    image: influxdb:1.8
    ports:
      - 8086:8086
    environment:
      - INFLUXDB_DB=powerapi_formula
      - INFLUXDB_HTTP_LOG_ENABLED=true
      - INFLUXDB_LOGGING_LEVEL=debug
      - INFLUXDB_HTTP_FLUX_ENABLED=true
      - DOCKER_INFLUXDB_INIT_ORG=org
    healthcheck:
      test: "curl -f http://localhost:8086/ping"
      interval: 3s
      timeout: 10s
      retries: 5
      start_period: 40s
    command: -config /etc/influxdb/influxdb.conf


  mongo:
    container_name: mongo
    image: mongo
    network_mode: "host"
    ports:
      - 27017:27017
    environment:
      - "TZ=Europe/Rome" 
    restart : always
    healthcheck:
      test: echo 'db.runCommand("ping").ok' | mongosh localhost:27017/ --quiet
      interval: 10s
      timeout: 10s
      retries: 5
      start_period: 40s

  power-api-hwpc:
    image: powerapi/hwpc-sensor:latest
    cap_add:
      - ALL
    network_mode: "host"
    privileged: true
    pid: "host"
    depends_on:
      mongo : 
        condition: service_healthy
    volumes:
      - /sys:/sys
      - /var/lib/docker/containers:/var/lib/docker/containers:ro
      - /tmp/powerapi-sensor-reporting:/reporting 
      - ./hwpc_config.json:/hwpc_config.json
    command:
      - "--config-file"
      - "/hwpc_config.json"
   

  power-api-smartwatts:
    image: powerapi/smartwatts-formula
    cap_add:
      - ALL
    volumes:
      - ./smartwatts_config.json:/smartwatts_config.json
    command:
      - "--config-file"
      - "/smartwatts_config.json"
    depends_on:
      influxdb : 
         condition : service_healthy
      power-api-hwpc : 
         condition : service_started
    network_mode: "host"
  ```

**Comment by roda82** [comment]
Are you sure that your docker compose is working ? Do you have ouputs on influxdb ? Can you share sensor and formula configuration files ?

**Comment by msalvati1997** [comment]
Yes, I'm sure it works but the latency expands with the use of architecture. 
Specifically, I am using Smartwatts to monitor the energy consumption of  functions running on docker containers in a distributed cloud environment.
What I do is start the powerapi architecture when a node in the distributed system starts running functions on some container. 

- Sensor conf file: 
```
{
  "name": "sensor",
  "verbose": false,
  "frequency": 500,
  "output": {
    "type": "mongodb",
    "uri": "mongodb://127.0.0.1",
    "database": "powerapi",
    "collection": "hwpcsensor"
  },
  "system": {
    "rapl": {
      "events": ["RAPL_ENERGY_PKG"],
      "monitoring_type": "MONITOR_ONE_CPU_PER_SOCKET"
    },
    "msr": {
      "events": ["TSC", "APERF", "MPERF"]
    }
  },
  "container": {
    "core": {
      "events": [
        "CPU_CLK_THREAD_UNHALTED:REF_P",
        "CPU_CLK_THREAD_UNHALTED:THREAD_P",
        "LLC_MISSES",
        "INSTRUCTIONS_RETIRED"
      ]
    }
  }
}
```
- formula conf file: 
```

{
  "verbose": false,
  "stream": true,
  "input": {
    "puller": {
      "type": "mongodb",
      "uri": "mongodb://127.0.0.1",
      "db": "powerapi",
      "collection": "hwpcsensor"
    }
  },
  "output": {
    "pusher_power": {
      "type": "influxdb",
      "model": "PowerReport",
      "uri": "127.0.0.1",
      "port": 8086,
      "db": "powerapi_formula",
      "collection": "smartwatts"
    }
  },
  "cpu-frequency-base": 2600,
  "cpu-frequency-min": 800,
  "cpu-frequency-max": 3500,
  "cpu-error-threshold": 2.0,
  "disable-dram-formula": true,
  "sensor-report-sampling-interval": 500
}
```

I use configuration scripts to generate the configuration files.

As the time the sensors are active increases, it seems that the latency expands and the power values are written to influxdb later and later. 

Thank you in advance.

**Comment by roda82** [comment]
Hello, 
Did you find a solution to the problem ? Otherwise, we need to modify chunck size used in the input and output databases in order to store values. 

Regards,

**Comment by msalvati1997** [comment]
My solution was to increase the frequency of hwpc sensor from 500 ms to 1000 ms! Also for sensor-report-sampling-interval This leads to dramatically decreased latencies!! Now it works well.

**Comment by gfieni** [comment]
Hello,
This performance regression is now fixed in the `2.1.0` version of SmartWatts.
Please open a new issue if you encounter other performance problems.
